### ISSUE TITLE
createWriterToModify/writeText results in new PDF with no changes

### ISSUE THREAD

**Issue by JoshLaseter** [galkahana/HummusJS#76]
(Background: I have no experience with the PDF format, so I may be missing something obvious)

I'm starting from the  ModifyExistingPageContent test, which works fine with the provided PDFs, but doesn't work with some of the PDFs I'm trying to modify. The code does seem to work for others, however, so I'm not sure what's wrong. 

```
var hummus = require('hummus');

function modifyFile(filename)
{
	console.log(filename);
	var pdfWriter = hummus.createWriterToModify('../input-files/' + filename, {
			modifiedFilePath: '../output-files/hummus-' + filename
		});
	
	var pageModifier = new hummus.PDFPageModifier(pdfWriter,0);
	var fontCouri = pdfWriter.getFontForFile('../input-files/Couri.ttf');
	pageModifier.startContext().getContext().writeText(
				'Test Text',
				75, 75,
				{
					font:fontCouri,
					size:50,
					colorspace:'gray',
					color:0x00
				}
			);
		
	pageModifier.endContext().writePage();
	pdfWriter.end();
}

console.log("============Working============");
modifyFile('BasicJPGImagesTest.PDF');
modifyFile('BasicTIFFImagesTest.PDF');
modifyFile('jrr-21-149.pdf');

console.log("============Non-Working============");
modifyFile('x-8-6-1187.pdf');
modifyFile('30-9-398.pdf');
```

What did strike me as strange is that the size of the Test Text is very different on the test PDFs supplied with hummus vs the one working example I have are significantly different. I'm guessing there is something intrinsically different about the working and non-working PDFs, so I wrote some quick code to use the parsing functions to see if anything jumped out at me (completely just fishing):

```
var hummus = require('hummus');

function writeInfo(filename) {
	console.log(filename);
	var pdfReader = hummus.createReader('../input-files/' + filename);
	console.log('encrypted  ', pdfReader.isEncrypted());
	console.log('PDF Level  ', pdfReader.getPDFLevel());
	console.log('Page Count ', pdfReader.getPagesCount());
	console.log('Obj Count  ', pdfReader.getObjectsCount());
}

console.log("============Working============");
writeInfo('BasicJPGImagesTest.PDF');
writeInfo('BasicTIFFImagesTest.PDF');
writeInfo('jrr-21-149.pdf');

console.log("============Non-Working============");
writeInfo('x-8-6-1187.pdf');
writeInfo('30-9-398.pdf');
```

Which prints:

```
============Working============
BasicJPGImagesTest.PDF
encrypted   false
PDF Level   1.3
Page Count  1
Obj Count   17
BasicTIFFImagesTest.PDF
encrypted   false
PDF Level   1.3
Page Count  45
Obj Count   420
jrr-21-149.pdf
encrypted   false
PDF Level   1.4
Page Count  8
Obj Count   70
============Non-Working============
x-8-6-1187.pdf
encrypted   false
PDF Level   1.2
Page Count  14
Obj Count   98
30-9-398.pdf
encrypted   false
PDF Level   1.2
Page Count  11
Obj Count   82
```

What jumps out at me here is that the failing pdfs are PDF Level 1.2 and the working ones are 1.3/1.4 -- this is a very small sample size, so I'm not positive that's meaningful. Also, again being new to PDF format, I don't actually know what the distinction between the levels is.

Any suggestions for a work-around? I have attached one working and one non-working sample PDF that I'm using if that's helpful. 

Thanks in advance!

[x-8-6-1187.pdf](https://github.com/galkahana/HummusJS/files/318846/x-8-6-1187.pdf)
[jrr-21-149.pdf](https://github.com/galkahana/HummusJS/files/318845/jrr-21-149.pdf)

**Comment by galkahana** [comment]
Hi,
i'll have to look into this some time. nothing comes to mind...but 1.2 pdfs...those are really old school, so it might be something i didn't get to test on and so may fail.
createWriterToModify should come with a log option, how about you hook a logfile too it, see if it sais something that can hint of the problem?

Gal.

**Comment by JoshLaseter** [comment]
Thanks for the quick reply.

I'm having trouble getting a log hooked up. I didn't see any examples of hooking up a logfile, but looking at the source I thought this was the right syntax:

```
	var pdfWriter = hummus.createWriterToModify('../input-files/' + filename, {
			modifiedFilePath: '../output-files/hummus-' + filename,
			log: '../output-files/hummus-' + filename + '.txt'
		});
```

I even tried simplifying it:

```
	var pdfWriter = hummus.createWriterToModify('../input-files/' + filename, {
			modifiedFilePath: '../output-files/hummus-' + filename,
			log: 'hummus.txt'
		});
```

but I'm still not getting any logs to show up anywhere. Any thoughts?

**Comment by galkahana** [comment]
nope.
I'm afraid that i'm frightfully busy in the coming two weeks, so that my ability to assist here is limited.
If I had to guess sounds like the problem is with what may be really old PDF files, or an old producer (the 1.2's). When i'm done with my current efforts I can look into it.
Till then, debugging will require some C++ skills, so if you got them head over to https://github.com/galkahana/PDF-Writer
OR
try to see if you can convert the 1.2s to a later version of PDF, or just limit their usage.

Gal.

**Comment by JoshLaseter** [comment]
Thanks Gal, understood and appreciate your help. I'll keep exploring and let you know if I discover anything useful.

**Comment by JoshLaseter** [comment]
I'm not sure if I should close this issue because I think there might still something to be addressed with the PDFPageModifier, but I'm going to anyways since I'm no longer seeking help: 

I managed a work-around for my particular situation. I found that if I created a new page and did the text manipulation *before* merging the existing page in, I was able to get the text to appear even on version 1.2 PDFs (interestingly, I think all of my resultant PDFs, 1.4s included, came out as 1.3s). If I swapped the order, however, it was still having issues, so I'm unclear if the problem is PDFPageModifier or writing text or some other general limitation within the older PDF specs.

Here's my currently working code for anyone else that's struggling on this point in the future (along with some additional scope creep from the original sample above to copy mediaBox size and work for multiple pages instead of just one):

```
function modifyFile(filename)
{
	console.log(filename);
	var pdfReader = hummus.createReader('../input-files/' + filename);
	var mediaBox = pdfReader.parsePage(0).getMediaBox();
	mediaBox[3] = mediaBox[3] + 35;
	var pageCount = pdfReader.getPagesCount();
	
	var pdfWriter = hummus.createWriter('../output-files/hummus/merge-pdf-pages/' + filename);
	var fontCouri = pdfWriter.getFontForFile('../input-files/Couri.ttf');
	
	for(pageIndex = 0; pageIndex < pageCount; pageIndex++)
	{	
		console.log('Write Page Index ' + pageIndex);
		var page = pdfWriter.createPage();
		page.mediaBox = mediaBox;
		
		var contentContext = pdfWriter.startPageContentContext(page);
	
		contentContext
				.BT() // Begin Text
				.k(0,0,0,1) // Set Color (CMYK, 0-1)
				.Tf(fontCouri,15) // Text font? Font & size?
				.Tm(1,0,0,1,5,20) // Text Matrix, set to coord 5,20
				.Tj('Test Text') // Show text
				.Tm(1,0,0,1,5,5) // Text Matrix, set to coord 5,5 (to place below previous line)
				.Tj('Page Index ' + pageIndex) // More text
				.ET() // End Text
				.q() // Push Current Matrix
				.cm(1,0,0,1,0,35);  // Set Current Matrix - scale to 100% (x and y), translate 0,35

		pdfWriter.mergePDFPagesToPage(page,
					'../input-files/' + filename,
					{type:hummus.eRangeTypeSpecific,specificRanges:[[pageIndex,pageIndex]]});
					
		contentContext.Q(); // Pop Current Matrix				

		pdfWriter.writePage(page);
	}	
	
	pdfWriter.end();
}
```

I'm sadly still stuck on input from (non-file) streams so I may be asking for help on that front in a different thread shortly (still trying to read all the documentation / samples / relevant cpp code first to make sure I'm not missing anything obvious).

Thanks again for the help (and excellent module)
